src/nvim/os/keyboard.c

```diff
@@ -0,0 +1,285 @@
+
+  if (!os_xlib_dlsym("XOpenDisplay", (void **)&_XOpenDisplay, err)
+      || !os_xlib_dlsym("XCloseDisplay", (void **)&_XCloseDisplay, err)
+      || !os_xlib_dlsym("XkbGetState", (void **)&_XkbGetState, err)) {
+    return false;
+  }
+  return xlib_opened;
+}
+
+/// Get information about Num/Caps/Scroll Lock state in Linux
+///
+/// To be used in os_mods_status() function.
+///
+/// @param[out]  mods  Holds the information about the status of various locks
+///                    in form of a pointer to integer bitmask.
+/// @param[out]  err  Location where error message is to be saved
+///
+/// @return -1 in case of error and a mask specifying which values
+///         in mods are valid otherwise.
+static ModMask os_get_locks_status(ModMask *const mods, Error *const err)
+{
+  if (!os_xlib_init(err)) {
+    return -1;
+  }
+
+  _XDisplay *const display = _XOpenDisplay(NULL);
+  if (display == NULL) {
+    api_set_error(err, kErrorTypeException,
+                  "Unable to open the display");
+    return -1;
+  }
+
+  _XkbStateRec xkb_state;
+  const _XStatus status = _XkbGetState(display, _XkbUseCoreKbd, &xkb_state);
+  if (status) {
+    api_set_error(err, kErrorTypeException,
+                  "Unable to get keyboard state of the display");
+
+    _XCloseDisplay(display);
+  }
+
+  int state = xkb_state.data[kLockedModsOffset];
+  if (state & kXNumLock) {
+    *mods |= kNumLock;
+  }
+
+  if (state & kXCapsLock) {
+    *mods |= kCapsLock;
+  }
+
+  if (state & kXScrollLock) {
+    *mods |= kScrollLock;
+  }
+
+  _XCloseDisplay(display);
+  return kCapsLock | kNumLock | kScrollLock;
+}
+
+#elif defined(WIN32)
+/// Get information about Num/Caps/Scroll Lock state in Windows.
+///
+/// To be used in os_mods_status() function.
+///
+/// @param[out]  mods  Holds the information about the status of various locks
+///                    in form of a pointer to integer bitmask.
+/// @param[out]  err  Location where error message is to be saved
+///
+/// @return -1 in case of error and a mask specifying which values
+///         in mods are valid otherwise.
+static ModMask os_get_locks_status(ModMask *const mods, Error *const err)
+{
+  *mods = 0;
+  if (GetKeyState(VK_CAPITAL) & 0x0001) {
+    *mods |= kCapsLock;
+  }
+  if (GetKeyState(VK_NUMLOCK) & 0x0001) {
+    *mods |= kNumLock;
+  }
+  if (GetKeyState(VK_SCROLL) & 0x0001) {
+    *mods |= kScrollLock;
+  }
+  return kCapsLock | kNumLock | kScrollLock;
+}
+
+#elif defined(__APPLE__)
+/// Get information about Num/Caps Lock state in MacOS
+///
+/// To be used in os_mods_status() function.
+///
+/// @param[out]  mods  Holds the information about the status of various locks
+///                    in form of a pointer to integer bitmask.
+/// @param[out]  err  Location where error message is to be saved
+///
+/// @return -1 in case of error and a mask specifying which values
+///         in mods are valid otherwise.
+static ModMask os_get_locks_status(ModMask *const mods, Error *const err)
```

Comment:
Would be a good idea to add `FUNC_ATTR_NONNULL_ALL FUNC_ATTR_WARN_UNUSED_RESULT` to all these functions.